File: src/deltavfs/injector.nim
import os
import osproc
import strutils
import strformat

import ../payload

import stew/endians2
import stew/byteutils

import winim/lean
from winim/inc/psapi import
  EnumProcessModules,
  GetModuleBaseNameA

proc remoteLoadLib(procHandle: HANDLE, moduleName: string): int = 
  var cModuleName = $moduleName

  echo GetLastError()

  let moduleNameAddr = VirtualAllocEx(
    procHandle,
    NULL,
    cast[SIZE_T](len(moduleName) + 1),
    MEM_COMMIT,
    PAGE_EXECUTE_READWRITE
  )
  echo GetLastError()

  WriteProcessMemory(procHandle, moduleNameAddr, LPCVOID(addr cModuleName[0]), SIZE_T(len(moduleName) + 1), NULL)

  echo GetLastError()

  echo &"Wrote to {toHex(cast[int](moduleNameAddr))}"

  let loadLibraryAddr = cast[int](LoadLibraryA)
  let resultAddr = VirtualAllocEx(
    procHandle,
    NULL,
    cast[SIZE_T](sizeof(HMODULE)),
    MEM_COMMIT,
    PAGE_READWRITE
  )

  echo GetLastError()

  let moduleNameShc = swapBytes(cast[uint64](moduleNameAddr))
  let loadLibraryShc = swapBytes(cast[uint64](loadLibraryAddr))
  let resultShc = swapBytes(cast[uint64](resultAddr))

  echo &"LoadLibraryA: {toHex(cast[int](loadLibraryAddr))}"
  echo &"resultAddr  : {toHex(cast[int](resultAddr))}"

  let payload = &"""
  0x53                 ; push rbx
  0x48 0x89 0xE3             ; mov rbp, rsp
  0x48 0x83 0xEC 0x20          ; sub rsp, 0x20
  0x48 0xB9 {toHex(moduleNameShc)}   ; mov rcx, <library name>
  0x48 0xBA {toHex(loadLibraryShc)}  ; mov rdx, <addr of LoadLibrary>
  0xFF 0xD2              ; call rdx
  0x48 0xBA {toHex(resultShc)}     ; movabs rdx, <result ptr>
  0x48 0x89 0x02             ; mov [rdx], rax
  0x48 0x89 0xDC             ; pop rsp, rbx
  0x5B                 ; pop rbx
  0xC3                 ; ret
  """

  echo payload

  # Some quick formatting to remove comments, additional whitespace, and newlines.
  var resultShellcode = newSeq[byte]()

  for line in splitLines(payload):
    let data = split(line, ";")[0]
      .strip()
      .replace("0x", "")
      .replace(" ", "")
      .hexToSeqByte()

    resultShellcode = resultShellcode & data

  echo resultShellcode

  # Allocate space in executable memory and write the shellcode.
  let shellcodeAddr = VirtualAllocEx(
    procHandle,
    NULL,
    len(resultShellcode),
    MEM_COMMIT,
    PAGE_EXECUTE_READWRITE
  )
  WriteProcessMemory(procHandle, shellcodeAddr, LPCVOID(addr resultShellcode[0]), SIZE_T(len(resultShellcode)), NULL)

  echo &"shellcode written to {toHex(cast[int](shellcodeAddr))}"

  let loaderThread = CreateRemoteThread(
    procHandle,
    NULL,
    0,
    cast[LPTHREAD_START_ROUTINE](shellcodeAddr),
    NULL,
    0,
    NULL
  )

  result = int(0)
  WaitForSingleObject(loaderThread, INFINITE)
  CloseHandle(loaderThread)

  ReadProcessMemory(procHandle, resultAddr, cast[LPVOID](addr result), 8, NULL)

  # Cleanup allocated memory.
  VirtualFree(moduleNameAddr, 0, MEM_RELEASE)
  VirtualFree(resultAddr, 0, MEM_RELEASE)
  VirtualFree(shellcodeAddr, 0, MEM_RELEASE)

proc getLibFuncOffset(moduleName: string, funcName: string): int =
  var cModuleName = $moduleName
  var cFuncName = &funcName

  # Load the module into memory, taking care not to resolve any dependent libraries.
  let moduleHandle = LoadLibraryExA(cast[LPCSTR](addr cModuleName[0]), HANDLE(0), DWORD(DONT_RESOLVE_DLL_REFERENCES))
  let moduleBaseAddr = cast[int](moduleHandle)

  # Determine the local address of the target function, by name.
  let localAddr = GetProcAddress(moduleHandle, cast[LPCSTR](cFuncName))

  echo &"moduleBaseAddr: {toHex(cast[int](moduleBaseAddr))}"
  echo &"localAddr:    {toHex(cast[int](localAddr))}"

  # Compute the delta offset between the module base and the function.
  let localDelta = cast[int](localAddr) - cast[int](moduleBaseAddr)

  localDelta

proc escalatePriv(procHandle: HANDLE) =
  var procToken: HANDLE
  var uid: LUID

  var tokenPrivs = create(TOKEN_PRIVILEGES)

  OpenProcessToken(procHandle, TOKEN_ADJUST_PRIVILEGES or TOKEN_QUERY, addr procToken)
  LookupPrivilegeValue(NULL, SE_DEBUG_NAME, addr uid)

  tokenPrivs.PrivilegeCount = 1
  tokenPrivs.Privileges[0].Luid = uid
  tokenPrivs.Privileges[0].Attributes = SE_PRIVILEGE_ENABLED

  AdjustTokenPrivileges(procToken, FALSE, tokenPrivs, DWORD(sizeof(tokenPrivs)), NULL, NULL)
  CloseHandle(procToken)

proc injectDll*(into: string) = 
  let binDir = currentSourcePath().splitPath().head / "../../bin"

  let capstoneDll = absolutePath(binDir / "capstone.dll")
  let payloadDll = absolutePath(binDir / "payload.dll")

  echo capstoneDll
  echo payloadDll

  # Start the target process in a suspended state.
  let target = startProcess(into, options={poEvalCommand})
  suspend(target)

  let targetHandle = OpenProcess(
    PROCESS_ALL_ACCESS,
    FALSE,
    DWORD(target.processId)
  )

  # Escalate SE_DEBUG privileges. Only seems to be useful when running
  # within an administrator process.
  # escalatePriv(targetHandle)

  resume(target)

  discard readLine(stdin)

  discard remoteLoadLib(targetHandle, capstoneDll)
  let payloadBase = remoteLoadLib(targetHandle, payloadDll)

  echo &"payload base address is {toHex(payloadBase)}"

  let localDelta = getLibFuncOffset(payloadDll, "mainThread")
  echo &"local delta: {toHex(localDelta)}"

  let remoteFunc = payloadBase + localDelta

  echo &"target function: {toHex(remoteFunc)}"

  discard readLine(stdin)

  CreateRemoteThread(
    targetHandle,
    NULL,
    0,
    cast[LPTHREAD_START_ROUTINE](cast[uint](remoteFunc)),
    NULL,
    0,
    NULL
  )

  # echo &"spawned mainThread in process {target.processId} with execution at {toHex(cast[int](mainThread))}"

  # Allocate and copy dllPath into the target process.
  # let remoteDllPath = VirtualAllocEx(
  #   targetHandle, 
  #   NULL, 
  #   cast[SIZE_T](len(dllPath)),
  #   MEM_COMMIT, 
  #   PAGE_EXECUTE_READ_WRITE
  # )

  # echo toHex(cast[uint](remoteDllPath))
  # echo toHex(cast[uint](LoadLibraryA))

  # WriteProcessMemory(
  #   targetHandle, 
  #   remoteDllPath, 
  #   cstring(dllPath), 
  #   cast[SIZE_T](len(dllPath)), 
  #   NULL
  # )

  # # Spawn a thread to execute LoadLibraryW with the dll
  # let loadThread = CreateRemoteThread(
  #   targetHandle,
  #   NULL,
  #   0,
  #   cast[LPTHREAD_START_ROUTINE](cast[uint](LoadLibraryA)),
  #   remoteDllPath,
  #   0,
  #   NULL
  # )

   #  if loadThread != HANDLE(0):
   #    CloseHandle(loadThread)

File: crates/deltavfs-inject/src/injector.rs
use shared::pe;

use std::ffi::c_void;
use std::io::{BufRead, BufReader};
use std::mem::size_of;
use std::process::{Command, Stdio};
use std::{ptr, thread};
use std::os::windows::process::CommandExt;

use anyhow::{anyhow, Result};
use hex;
use log::*;

use windows::Win32::Foundation::{CloseHandle, GetLastError, HANDLE, HINSTANCE};
use windows::Win32::System::Diagnostics::Debug::{ReadProcessMemory, WriteProcessMemory};
use windows::Win32::System::LibraryLoader::{
    GetProcAddress, LoadLibraryExA, DONT_RESOLVE_DLL_REFERENCES,
};
use windows::Win32::System::Memory::{VirtualFree, MEM_COMMIT, MEM_FREE, PAGE_EXECUTE_READWRITE};
use windows::Win32::System::Threading::{CREATE_SUSPENDED, CreateRemoteThread, WaitForSingleObject};
use windows::Win32::System::{
    Memory::VirtualAllocEx,
    Threading::{OpenProcess, PROCESS_ALL_ACCESS},
};

pub unsafe fn inject_into(exec_path: &str) -> Result<()> {
    info!(
        "Preparing to inject payload into executable at '{}'",
        exec_path
    );

    // Spawn the process in a suspended state.
    let mut proc = Command::new(exec_path)
        .creation_flags(CREATE_SUSPENDED)
        .stdout(Stdio::piped())
        .spawn()?;

    let stdout = proc
        .stdout
        .take()
        .expect("Failed to open stdout of child process.");

    // Open threaded reader to echo child stdout to parent.
    thread::spawn(move || {
        for line in BufReader::new(stdout).lines() {
            println!("[~] {}", line.unwrap());
        }
    });

    let proc_handle = OpenProcess(PROCESS_ALL_ACCESS, false, proc.id());

    let library = "C:\\Users\\green\\Dev\\rust\\deltavfs\\target\\debug\\deltavfs_hook.dll";
    let lib_base = remote_loadlib(proc_handle, library)?;

    // Determine the location of the mainThread function within the target.
    let local_delta = get_lib_offset(library, "hook_init\0")?;
    let entry_point = lib_base + local_delta;

    println!("Found remote entry point at: {:x?}", entry_point);

    // Create a thread at mainThread to begin hook installation.
    CreateRemoteThread(
        proc_handle,
        ptr::null_mut(),
        0,
        std::mem::transmute(entry_point),
        ptr::null_mut(),
        0,
        ptr::null_mut(),
    );

    loop {

    }

    Ok(())
}

unsafe fn remote_loadlib(proc_handle: HANDLE, library: &str) -> Result<usize> {
    // Allocate space in the remote process to store the path of the library.
    let lib_path_ptr = VirtualAllocEx(
        proc_handle,
        ptr::null_mut(),
        library.len() + 1,
        MEM_COMMIT,
        PAGE_EXECUTE_READWRITE,
    );

    println!("{:x?}", lib_path_ptr);

    println!(
        "Remote allocated {} bytes for library path.",
        library.len() + 1
    );

    // Allocate space in the remote process to store the result of our LoadLibraryA call.
    let result_ptr = VirtualAllocEx(
        proc_handle,
        ptr::null_mut(),
        size_of::<HINSTANCE>(),
        MEM_COMMIT,
        PAGE_EXECUTE_READWRITE,
    );

    println!(
        "Remote allocated {} bytes for LoadLibraryA result.",
        size_of::<HINSTANCE>()
    );

    // Write the path of the library to memory.
    WriteProcessMemory(
        proc_handle,
        lib_path_ptr,
        library.as_bytes().as_ptr() as _,
        library.len(),
        ptr::null_mut(),
    );

    let loadlib_ptr = pe::get_func_addr("kernel32", "LoadLibraryA")?;

    // Create the payload string, embed relevant addresses, and convert to byte array.
    let payload_str = format!(
        "      
        0x53
        0x48 0x89 0xE3
        0x48 0x83 0xEC 0x20
        0x48 0xB9 {}
        0x48 0xBA {}
        0xFF 0xD2
        0x48 0xBA {}
        0x48 0x89 0x02
        0x48 0x89 0xDC
        0x5B
        0xC3                                
    ",
        to_hex(lib_path_ptr),
        to_hex(loadlib_ptr),
        to_hex(result_ptr)
    )
    .replace(" ", "")
    .replace("\n", "")
    .replace("0x", "");

    let mut payload = hex::decode(payload_str).unwrap();

    println!("{:X?}", payload);

    // Allocate remote memory for shellcode.
    let shellcode_ptr = VirtualAllocEx(
        proc_handle,
        ptr::null_mut(),
        payload.len(),
        MEM_COMMIT,
        PAGE_EXECUTE_READWRITE,
    );

    // Write shellcode into process.
    let mut bytes_written: usize = 0;
    WriteProcessMemory(
        proc_handle,
        shellcode_ptr,
        payload.as_mut_ptr() as _,
        payload.len(),
        &mut bytes_written,
    );

    println!("Remote LoadLibrary shellcode written to remote process.");
    println!("shellcode ptr: {:x?}", shellcode_ptr);
    println!("result ptr: {:x?}", result_ptr);

    // Spawn a remote thread to execute the payload.
    let loader_thread = CreateRemoteThread(
        proc_handle,
        ptr::null_mut(),
        0,
        Some(std::mem::transmute::<
            _,
            unsafe extern "system" fn(lpthreadparameter: *mut c_void) -> u32,
        >(shellcode_ptr)),
        ptr::null_mut(),
        0,
        ptr::null_mut(),
    );

    println!("Thread spawned in remote process.");

    WaitForSingleObject(loader_thread, 100000000);
    CloseHandle(loader_thread);

    // Read the remote process' memory for the result.
    let mut result = HINSTANCE::default();
    let mut bytes_read: usize = 0;
    ReadProcessMemory(
        proc_handle,
        result_ptr,
        &mut result as *mut HINSTANCE as _,
        size_of::<HINSTANCE>(),
        &mut bytes_read,
    );

    if bytes_read != size_of::<HINSTANCE>() {
        return Err(anyhow!(
            "Read an invalid number of bytes. Expected {}, got {}.",
            size_of::<HINSTANCE>(),
            bytes_read
        ));
    }

    println!("{:x?}", result_ptr);

    // Cleanup remote memory.
    VirtualFree(lib_path_ptr, 0, MEM_FREE);
    VirtualFree(result_ptr, 0, MEM_FREE);
    VirtualFree(shellcode_ptr, 0, MEM_FREE);

    Ok(std::mem::transmute::<_, usize>(result))
}

unsafe fn get_lib_offset(library: &str, func_name: &str) -> Result<usize> {
    let module_handle = LoadLibraryExA(library, HANDLE(0), DONT_RESOLVE_DLL_REFERENCES);

    let module_base = std::mem::transmute::<_, *const c_void>(module_handle);

    let local_ptr = match GetProcAddress(module_handle, func_name) {
        Some(x) => x as *const () as *const c_void,
        None => return Err(anyhow!("Unable to find location of func.")),
    };

    let err = GetLastError();
    println!("{:?}", err);

    Ok((local_ptr as usize) - (module_base as usize))
}

fn to_hex(ptr: *const c_void) -> String {
    hex::encode((ptr as usize).to_le_bytes())
}
